### PR TITLE
Working with server data

### PR DESCRIPTION
**Pull request by plantysnake**

### CODE REVIEW

**Comment by plantysnake** [comment]
- [x] Display table information on trials, returning mock json data from a service
- [x] Return the data from a node server
- [ ] Use local styles with sass
- [ ] Extract global sass style and variables
- [x] Save data to local component state
- [ ] Introduce global state with redux
- [ ] Implement data loading with redux saga

**Comment by plantysnake** [comment]
https://github.com/plantysnake/TestingKnowledge-api

File: src/js/components/App.jsx [hunk 1]
```diff
@@ -1,12 +1,97 @@
...
+}
+
+Table.propTypes = {
+    trId: PropTypes.number,
+    title: PropTypes.string,
+    disease: PropTypes.string,
+    phase: PropTypes.string
+}
+
+const style = {
```
Comment (marina-igorevna-mishur) [inline src/js/components/App.jsx]: please move to separate `Table.scss` and `Row.scss` files

File: src/js/services/AppService.jsx [hunk 2]
```diff
@@ -1,5 +1,8 @@
...
+import { trials } from "utils/API"
+
 function getData() {
-	return "Hello World?"
+    return http.read(trials)
 }
 
-export default { getData }
\ No newline at end of file
+export { getData }
```
Comment (marina-igorevna-mishur) [inline src/js/services/AppService.jsx]: this is not a `.jsx` file, please change the extension

Comment (marina-igorevna-mishur) [reply src/js/services/AppService.jsx]: please change to a more meaningful name

File: tests/App.test.jsx [hunk 3]
```diff
@@ -1,24 +1,33 @@
...
+import React from "react"
+import Enzyme, { shallow } from "enzyme"
+import App from "components/App.jsx"
+import AppService from "services/AppService.jsx"
+import Adapter from "enzyme-adapter-react-16"
 
-const sinon = require('sinon');
+const sinon = require("sinon")
 
 Enzyme.configure({ adapter: new Adapter() })
```
Comment (marina-igorevna-mishur) [inline tests/App.test.jsx]: you could add a jest setup file to configure adapter for all tests

File: tests/App.test.jsx [hunk 4]
```diff
@@ -1,24 +1,33 @@
...
 
-describe('App component', () => {
-    it('gets right string', () => {
-      const wrapper = shallow(<App />);
-      const text = wrapper.find('div').text();
-      expect(text).toEqual('Hello World?');
-    });
+describe("App component", () => {
```
Comment (marina-igorevna-mishur) [inline tests/App.test.jsx]: please separate and update tests

File: tests/App.test.jsx [hunk 5]
```diff
@@ -1,24 +1,33 @@
...
-import AppService from 'services/AppService.jsx';
-import Adapter from 'enzyme-adapter-react-16';
+import React from "react"
+import Enzyme, { shallow } from "enzyme"
+import App from "components/App.jsx"
+import AppService from "services/AppService.jsx"
+import Adapter from "enzyme-adapter-react-16"
 
-const sinon = require('sinon');
+const sinon = require("sinon")
```
Comment (marina-igorevna-mishur) [inline tests/App.test.jsx]: last time I checked `import sinon from "sinon"` was working for me here

File: tests/AppService.test.jsx [hunk 6]
```diff
@@ -1,9 +1,7 @@
...
-import { getData } from 'services/AppService.jsx';
+import { getData } from "services/AppService.jsx"
 
-describe('testing primitive', () => {
-    it('returns hello world"?"', () => {
-        expect(getData()).toBe("Hello World?");
-    });
-});
\ No newline at end of file
+describe("testing primitive", () => {
```
Comment (marina-igorevna-mishur) [inline tests/AppService.test.jsx]: please update tests

File: webpack.config.js [hunk 7]
```diff
@@ -1,49 +1,69 @@
...
-    }
-  },
-  module: {
-    rules: [
-      {
-        test: /\.(js|jsx)$/,
-        exclude: /node_modules/,
-        use: {
-          loader: "babel-loader"
+    entry: "./src/index.js",
```
Comment (marina-igorevna-mishur) [inline webpack.config.js]: would make sense to use `path.resolve` here too

File: tests/AppService.test.jsx [hunk 8]
```diff
@@ -1,9 +1,7 @@
-import React from 'react';
-import ReactDOM from 'react-dom';
-import { getData } from 'services/AppService.jsx';
+import { getData } from "services/AppService.jsx"
```
Comment (marina-igorevna-mishur) [inline tests/AppService.test.jsx]: neither `AppService` nor `AppService.test` are `.jsx`

File: .vscode/launch.json [hunk 9]
```diff
@@ -47,6 +48,25 @@
             "env": {
                 "NODE_ENV": "unit"
             }
+        },
+        {
+            "type": "node",
+            "request": "launch",
+            "name": "Eslint Current File",
```
[the remark this answers is not in the document]
Comment (plantysnake) [reply .vscode/launch.json]: Fixed in sagas branch

File: src/js/components/App.js.old [hunk 10]
```diff
@@ -2,10 +2,10 @@ import React from "react"
 import "../../css/style.css"
 
 export default class App extends React.Component {
```
[the remark this answers is not in the document]
Comment (plantysnake) [reply src/js/components/App.js.old]: Will be removed

File: src/js/components/App.jsx [hunk 11]
```diff
@@ -1,12 +1,97 @@
...
+    constructor(props) {
+        super(props)
+        this.state = {
+            head: [],
+            trials: []
+        }
+    }
+
+    componentDidMount() {
+        this.getTable()
```
[the remark this answers is not in the document]
Comment (plantysnake) [reply src/js/components/App.jsx]: Already moved

File: src/js/components/App.jsx (hunk 1 above)
Comment (plantysnake) [reply src/js/components/App.jsx]: Already

File: src/js/components/__snapshots__/App.test.js.snap [hunk 12]
```diff
@@ -0,0 +1,3 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`<App /> should render a base layout 1`] = `ShallowWrapper {}`;
```
Comment (marina-igorevna-mishur) [inline src/js/components/__snapshots__/App.test.js.snap]: this should not be an empty wrapper, should be `<Table/>`, see https://stackoverflow.com/questions/54419342/jest-enzyme-shallowwrapper-is-empty-when-creating-snapshot

File: src/js/services/TableDataService.test.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+            {
+                "id": 225263,
+                "title": "Test",
+                "disease": "NA: ClinicalTrials.gov",
+                "phase": "IV"
+            }
+        ]
+    }
+
+    it("returns correct data", () => {
```
Comment (marina-igorevna-mishur) [inline src/js/services/TableDataService.test.js]: to test the service you should stub `http.read` and `JSON.parse`, and check that:
- `http.read` was called with the `trials` argument,
- `JSON.parse` was called with the string `http.read` returned

File: src/js/components/Table.test.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
+import React from "react"
+import { Table } from "components/Table.jsx"
+import sinon from "sinon"
+
+describe("<Table />", () => {
+    let component
+
+    const columns = ["Trial id", "Title", "Disease", "Trial phase"]
+    const trials = [
```
[the remark this answers is not in the document]
Comment (plantysnake) [reply src/js/components/Table.test.js]: Already using, will be extracted as test data

File: src/js/components/Table.test.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    ]
+
+    beforeEach(() => {
+        const loadColumnsSpy = sinon.spy()
+        const loadTrialsSpy = sinon.spy()
+
+        component = shallow(<Table
+            head={columns}
+            trials={trials}
+            loadColumns={loadColumnsSpy}
```
[the remark this answers is not in the document]
Comment (plantysnake) [reply src/js/components/Table.test.js]: Yeah, I'll re-edit this,

File: src/js/components/Table.jsx [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        })
+    }
+
+    loadTableData = () => {
+        getTrialsInfo().then(this.onTableDataLoaded)
+    }
+
+    renderTableColumns = () => {
+        return this.state.head.map(column => {
+            return <td key={column} >{column}</td>
```
[the remark this answers is not in the document]
Comment (plantysnake) [reply src/js/components/Table.jsx]: Done

File: src/js/components/Table.jsx [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+        })
+    }
+
+    renderTableBody = () => {
+        const table = this.state.trials.map(function (trial) {
```
[the remark this answers is not in the document]
Comment (plantysnake) [reply src/js/components/Table.jsx]: Done lambda

File: src/js/components/__snapshots__/App.test.js.snap (hunk 12 above)
Comment (plantysnake) [reply src/js/components/__snapshots__/App.test.js.snap]: Fixed

File: src/js/components/Table.test.js [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+            loadTrials={loadTrialsSpy}
+        />)
+    })
+
+    afterEach(() => {
+        component.unmount()
```
[the remark this answers is not in the document]
Comment (plantysnake) [reply src/js/components/Table.test.js]: Done

File: src/js/services/TableDataService.test.js (hunk 13 above)
Comment (plantysnake) [reply src/js/services/TableDataService.test.js]: Done.

File: src/js/components/Table.test.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (plantysnake) [reply src/js/components/Table.test.js]: Re-done

File: src/js/services/TableDataService.test.js [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+        return TableDataService.getTrialsInfo().then(data => data)
+    })
+
+    afterEach(() => {
+        httpStub.restore()
+        JSONParseStub.restore()
+    })
+
+    it("http.read was called with the trials argument", () => {
+        sinon.assert.calledWith(httpStub, "http://127.0.0.1:3001/data.json")
```
Comment (marina-igorevna-mishur) [inline src/js/services/TableDataService.test.js]: why duplicate the constant here?

File: src/js/services/TableDataService.test.js [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+import sinon from "sinon"
+
+describe("testing service", () => {
+    let httpStub
+    let JSONParseStub
+
+    beforeEach(() => {
+        httpStub = sinon.stub(http, "read").returns(Promise.resolve("trials argument"))
+        JSONParseStub = sinon.stub(JSON, "parse")
+        return TableDataService.getTrialsInfo().then(data => data)
```
Comment (marina-igorevna-mishur) [inline src/js/services/TableDataService.test.js]: didn't test the result

File: src/js/services/TableDataService.test.js (hunk 19 above)
Comment (plantysnake) [reply src/js/services/TableDataService.test.js]: Imported trials from API

File: src/js/services/TableDataService.test.js (hunk 20 above)
Comment (plantysnake) [reply src/js/services/TableDataService.test.js]: Added result testing

**Pull request merged**